[Elixir] new concept exercise: `errors` (#1607)
exercism/v3: languages/elixir/exercises/concept/errors/.docs/introduction.md

Errors happen. In Elixir, while people often say to "let it crash", there are times when we need to rescue the function call to a known good state to fulfil a software contract. In some languages, errors are used as method of control flow, but in Elixir, this pattern is discouraged. We can often recognize functions that may rause an error just by their name: functions that raise errors are to have `!` at the end of their name. This is in comparison with functions that return `{:ok, value}` or `:error`. Look at these library examples:

```elixir
Map.fetch(%{a: 1}, :b)
# => :error
Map.fetch!(%{a: 1}, :b)
# => raises KeyError
```

## Try .. Rescue

Elixir provides a construct for rescuing from errors using `try .. rescue`

```elixir
try do                             #1
  raise RuntimeError, "error"      #2
rescue
  e in RuntimeError -> :error      #3
end
```

Let's examine this construct:

- **Line 1**, the block is declared with `try`

> - **Line 2**, the function call which may error is placed here, in this case we are calling `raise/1`

Reviewer: ```suggestion
- **Line 2**, the function call which may error is placed here, in this case we are calling `raise/2`
```

- **Line 3**, in the `rescue` section, we pattern match on the _Module_ name of the error raised
  - on the left side of `->`:
    - `e` is matched to the error struct
    - `in` is a keyword in this occurrence

>     - `RuntimeError` is the module to matched, but can match on any error module, or `_` all errors.

Reviewer: "module to matched" -> "module to be matched"?

I must say I'm struggling to understand the "but" part of this sentence.

> or `_` all errors

Sounds to me like there's at least one word missing here, but not sure which one... "or use `_` to match on all errors", is that what it was supposed to mean?

If the goal is to hint at this alternative:
```
rescue
  _ -> :error
end
```

Then I think that needs to be a sentence of its own, after the list. Because the list lists each word in the phrase `e in RuntimeError` as a separate item, but we're talking about a replacement for the whole phrase, not just the `RuntimeError` part.

If you choose to rephrase this, the same change should happen in `after.md`.

Author: ```suggestion
    - `RuntimeError` is the Error that we want to rescue.  If we wanted to rescue from all areas, we could use `_`.
```

> Errors happen. In Elixir, while people often say to "let it crash", there are times when we need to rescue the function call to a known good state to fulfil a software contract. In some languages, errors are used as method of control flow, but in Elixir, this pattern is discouraged. We can often recognize functions that may rause an error just by their name: functions that raise errors are to have `!` at the end of their name. This is in comparison with functions that return `{:ok, value}` or `:error`. Look at these library examples:

Reviewer: ```suggestion
Errors happen. In Elixir, while people often say to "let it crash", there are times when we need to rescue the function call to a known good state to fulfil a software contract. In some languages, errors are used as method of control flow, but in Elixir, this pattern is discouraged. We can often recognize functions that may raise an error just by their name: functions that raise errors are to have `!` at the end of their name. This is in comparison with functions that return `{:ok, value}` or `:error`. Look at these library examples:
```

Reviewer: I hope in `after.md` we could explain better what it means "to use errors for control flow" so that people know exactly what to avoid and why.

Reviewer: I'm not sure if GitHub showed well what my suggested change did. It was to fix the typo "rause" -> "raise", and the typo is still there.

```elixir
Map.fetch(%{a: 1}, :b)
# => :error
Map.fetch!(%{a: 1}, :b)
# => raises KeyError
```

## Try .. Rescue

Elixir provides a construct for rescuing from errors using `try .. rescue`

```elixir
try do                             #1
  raise "error"                    #2
rescue
  e in RuntimeError -> :error      #3
end
```

Let's examine this construct:

- **Line 1**, the block is declared with `try`
- **Line 2**, the function call which may error is placed here, in this case we are calling `raise/1`
- **Line 3**, in the `rescue` section, we pattern match on the _Module_ name of the error raised
  - on the left side of `->`:
    - `e` is matched to the error struct
    - `in` is a keyword in this occurrence

>     - `RuntimeError` is the module to matched, but can match on any error module, or `_` all errors.

Reviewer: People might find this surprising. I'm raising a string but rescuing a `RuntimeError` struct? It might need an explanation why this happens, or a change to a more explicit raise - `raise, RuntimeError, "error"`.
